designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -266,15 +269,34 @@ class DeploymentServiceImpl(dispatcher: DeploymentManagerDispatcher,
   }
 
   private def getStateFromEngine(deploymentManager: DeploymentManager, processIdWithName: ProcessIdWithName)
-                                (implicit ec: ExecutionContext, freshnessPolicy: DataFreshnessPolicy): Future[WithDataFreshnessStatus[Option[ProcessState]]] =
-    deploymentManager.getProcessState(processIdWithName.name).recover {
+                                (implicit ec: ExecutionContext, freshnessPolicy: DataFreshnessPolicy): Future[WithDataFreshnessStatus[Option[ProcessState]]] = {
+    lazy val failedToGetResult =
+      WithDataFreshnessStatus(
+        Option(SimpleProcessStateDefinitionManager.errorFailedToGet),
+        cached = false)
+
+    @volatile var statusFetchCompleted = false
+    val stateFromEngineFuture = deploymentManager.getProcessState(processIdWithName.name).recover {
       case NonFatal(e) =>
-        logger.warn(s"Failed to get status of ${processIdWithName}: ${e.getMessage}", e)
-        WithDataFreshnessStatus(
-          Some(SimpleProcessStateDefinitionManager.errorFailedToGet),
-          cached = false)
+        logger.warn(s"Failed to get status of ${processIdWithName.name}: ${e.getMessage}", e)
+        failedToGetResult
+    }.transform { result =>
+      statusFetchCompleted = true
+      result
     }
 
+    scenarioStateTimeout.map { timeout =>
+      Future.firstCompletedOf(Seq(
+        stateFromEngineFuture,
+        akka.pattern.after(timeout) {
+          if (!statusFetchCompleted) {
+            logger.warn(s"Timeout: $timeout occurred during waiting for response from engine for ${processIdWithName.name}")
+          }
+          Future.successful(failedToGetResult)
```

Comment:
```scala
if (!statusFetchCompleted) {
  logger.warn(s"Timeout: $timeout occurred during waiting for response from engine for ${processIdWithName.name}")
  Future.successful(failedToGetResult)
} else {
  stateFromEngineFuture
}
```

if `statusFetchCompleted` is true ,`stateFromEngineFuture` should be completed by now, so we may return it. WDYT?

Reply:
looks a lot better now :)